src/vsr.zig

```diff
@@ -1293,7 +1293,8 @@ const ViewChangeHeadersArray = struct {
         assert(headers.command == .start_view);
         // This function is only called by a replica that is lagging behind the primary's
         // checkpoint, so the start_view has a full suffix of headers.
-        assert(headers.array.get(0).op >= constants.journal_slot_count);
```

Comment:
Thinking out loud:

`start_view_into_do_view_change` is called by a replica that is transitioning from normal -> view_change, but is lagging behind the primary by at least a checkpoint, and the SV it has from the old view is from that newer checkpoint.

Using the checkpoint diagram for the purpose of mental model...:

```
///     commit log (ops)           │ write-ahead log (slots)
///     0   4   8   2   6   0   4  │ 0---4---
///   0 ─────✓·%                   │ 01234✓6%   initial log fill
///   1 ───────────✓·%             │ 890✓2%45   first wrap of log
///   2 ─────────────────✓·%       │ 6✓8%0123   second wrap of log
///   3 ───────────────────────✓·% │ 4%67890✓   third wrap of log
```

- In the first wrap, that means that our `self.op < 8`, so `headers.array.get(0).op >= 8`. `8 - 1` is our trigger, so it is the highest op that could possibly be in our log.
- So I think that the old assert was correct

Reply:
Got it! That `-1` in the computation of the frist checkpoint threw me into a loop. I was thinking we have some fancy logic where `op=0` doesn't actually go into the loog, because it's a root op or something (so, the first checkpoint is physically shorter than the rest). But that `-1` is just count <-> index conversion --- the first checkpoint has exactly `checkpoint_interval` ops (like all the other checkpoints), the first op number is 0, the last op number is `len - 1`